File: java/2015/day03b/Program.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.HashSet;

import aoc_utils.geometry.Position2D;

public class Program {
    public static void usage() {
        System.err.println("usage: java -jar Program.jar <input file>");
        System.exit(1);
    }

    public static int process(String filename) {
        HashSet<Position2D<Integer>> positions = new HashSet<Position2D<Integer>>();
        try ( BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            Position2D<Integer> santa = new Position2D<Integer>(0, 0);
            Position2D<Integer> roboSanta = new Position2D<Integer>(0, 0);
            boolean santaMoves = true;
            while ( (line = reader.readLine()) != null ) {
                for ( char ch : line.toCharArray() ) {
                    switch ( ch ) {
                        case '^':
                            if ( santaMoves ) {
                                santa.y += 1;
                            } else {
                                roboSanta.y += 1;
                            }
                            break;
                        case 'v':
                            if ( santaMoves ) {
                                santa.y -= 1;
                            } else {
                                roboSanta.y -= 1;
                            }
                            break;
                        case '<':
                            if ( santaMoves ) {
                                santa.x -= 1;
                            } else {
                                roboSanta.x -= 1;
                            }
                            break;
                        case '>':
                            if ( santaMoves ) {
                                santa.x += 1;
                            } else {
                                roboSanta.x += 1;
                            }
                            break;
                        default:
                            break;
                    }
                    if ( santaMoves ) {
                        positions.add(new Position2D<>(santa));
                    } else {
                        positions.add(new Position2D<>(roboSanta));
                    }
                    santaMoves = !santaMoves;
                }
            }
        } catch ( IOException e ) {
            System.err.println("Error reading file: " + e.getMessage());
            System.exit(1);
        }
        return positions.size();
    }

    public static final void main(String[] args) {
        if ( args.length < 1 ) {
            usage();
        }
        String filename = args[0];
        int result = process(filename);
        System.out.println("result = " + result);
    }
}

File: kotlin/2015/day03b/Program.kt
import java.io.File

data class Position2D(val x: Int, val y: Int)

fun usage() {
    System.err.println("usage: java -jar Program.jar <input file>")
    kotlin.system.exitProcess(1)
}

fun process(content: String) : Int {
    var santa = Position2D(0, 0)
    var roboSanta = Position2D(0, 0)
    var positions = mutableSetOf(santa)
    var santaMove = true
    var nextPosition: Position2D

    for ( ch in content ) {
        if ( santaMove ) {
            santa = when ( ch ) {
                '^' -> Position2D(santa.x, santa.y + 1)
                'v' -> Position2D(santa.x, santa.y - 1)
                '<' -> Position2D(santa.x - 1, santa.y)
                '>' -> Position2D(santa.x + 1, santa.y)
                else -> santa
            }
            nextPosition = santa
        } else {
            roboSanta = when ( ch ) {
                '^' -> Position2D(roboSanta.x, roboSanta.y + 1)
                'v' -> Position2D(roboSanta.x, roboSanta.y - 1)
                '<' -> Position2D(roboSanta.x - 1, roboSanta.y)
                '>' -> Position2D(roboSanta.x + 1, roboSanta.y)
                else -> roboSanta
            }
            nextPosition = roboSanta
        }
        positions.add(nextPosition)
        santaMove = !santaMove
    }

    return positions.count()
}

fun main(args: Array<String>) {
    if ( args.isEmpty()) {
        usage()
    }
    val filename = args[0]
    try {
        val content = File(filename).readText()
        val result = process(content)
        println("result = $result")
    } catch ( e: Exception ) {
        System.err.println("error reading file: ${e.message}")
    }
}
